host/sgx/sgx_enclave_common.c

```diff
@@ -0,0 +1,214 @@
+static void* _module;
+
+static void _unload_libsgx_enclave_common(void)
+{
+    if (_module)
+        oe_shared_library_unload(_module);
+}
+
+static void _load_libsgx_enclave_common_impl(void)
+{
+    oe_result_t result = OE_FAILURE;
+    void* module = _module;
+    if (!module)
+    {
+#if _WIN32
+        const char* library_name = "sgx_enclave_common.dll";
+#else
+        const char* library_name = "libsgx_enclave_common.so";
+#endif
+        OE_TRACE_INFO("Loading %s\n", library_name);
+        module = oe_shared_library_load(library_name);
+        if (module)
+        {
+            *(void**)&_enclave_create =
+                oe_shared_library_lookup(module, "enclave_create");
+            if (!_enclave_create)
+            {
+                OE_TRACE_ERROR("enclave_create function not found.\n");
+                goto done;
+            }
+
+            *(void**)&_enclave_load_data =
+                oe_shared_library_lookup(module, "enclave_load_data");
+            if (!_enclave_load_data)
+            {
+                OE_TRACE_ERROR("enclave_load_data function not found.\n");
+                goto done;
+            }
+
+            *(void**)&_enclave_initialize =
+                oe_shared_library_lookup(module, "enclave_initialize");
+            if (!_enclave_initialize)
+            {
+                OE_TRACE_ERROR("enclave_initialize function not found.\n");
+                goto done;
+            }
+
+            *(void**)&_enclave_delete =
+                oe_shared_library_lookup(module, "enclave_delete");
+            if (!_enclave_delete)
+            {
+                OE_TRACE_ERROR("enclave_delete function not found.\n");
+                goto done;
+            }
+
+            *(void**)&_enclave_set_information =
+                oe_shared_library_lookup(module, "enclave_set_information");
+            if (!_enclave_set_information)
+            {
+                OE_TRACE_ERROR("enclave_set_information function not found.\n");
+                goto done;
+            }
+
+            _module = module;
+            atexit(_unload_libsgx_enclave_common);
+            result = OE_OK;
+            OE_TRACE_INFO("Loaded %s\n", library_name);
+        }
+    }
+
+done:
+    if (result != OE_OK)
+    {
+        if (module)
+            oe_shared_library_unload(module);
+    }
+}
+
+static bool _load_libsgx_enclave_common(void)
+{
+    static oe_once_type _once;
+    oe_once(&_once, _load_libsgx_enclave_common_impl);
+    return (_module != NULL);
+}
+
+/****** Wrappers for functions in libsgx_enclave_common.so **************/
+void* COMM_API enclave_create(
+    COMM_IN_OPT void* base_address,
+    COMM_IN size_t virtual_size,
+    COMM_IN size_t initial_commit,
+    COMM_IN uint32_t type,
+    COMM_IN const void* info,
+    COMM_IN size_t info_size,
+    COMM_OUT_OPT uint32_t* enclave_error)
+{
+    if (!_load_libsgx_enclave_common())
+    {
+        *enclave_error = ENCLAVE_NOT_SUPPORTED;
```

Comment:
Can we make these wrappers part of the OE side instead? (i.e. `oe_` prefixed, return `oe_result_t` values etc.)

* I would prefer not to overload the meaning of SGX's error codes; the conditions for this are pretty specific on the SGX side today: SGX feature set or AESM doesn't exist. Here it's really an `OE_PLATFORM_ERROR` or a more unique error of your choice.
* I would prefer not to create the illusion this is actually a call into the sgx_enclave_common library already (i.e. folks may end up hunting for the function name in the SGX repo instead of OE based on an incorrect expectation of the error code).

Reply:
I have addressed this differently. See #3155 
The wrapper is thin and just forwards the call to SGX enclave common and does not overload the meaning of SGX error codes.
Failure to load sgx enclave common is treated as a catastrophic error and terminates the application. This is not that big of a deviation from the current behavior where the application won't load if sgx enclave common is not present in the system.